### PR TITLE
Setting <input type=file>.files

### PR DESCRIPTION
**Pull request by annevk**
For https://github.com/whatwg/html/pull/2866.

### CODE REVIEW

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/257217574)

# Firefox (nightly) #
Testing web-platform-tests at revision 465b82d6389b220a2c92e72b32fdc4df184208f1
Using browser at version BuildID 20170724100304; SourceStamp 5928d905c0bc0b28f5488b236444c7d7991cf8d4
Starting 10 test iterations
All results were stable
## All results ##
<details>
<summary>1 test ran</summary>
<details>
<summary><a href="http://w3c-test.org/submissions/6617/html/semantics/forms/the-input-element/files.html">/html/semantics/forms/the-input-element/files.html</a></summary>


|                Subtest                | Results |                                         Messages                                        |
|---------------------------------------|---------|-----------------------------------------------------------------------------------------|
|                                       | OK      |                                                                                         |
| `files for input type=hidden`         | PASS    |                                                                                         |
| `files for input type=text`           | PASS    |                                                                                         |
| `files for input type=search`         | PASS    |                                                                                         |
| `files for input type=tel`            | PASS    |                                                                                         |
| `files for input type=url`            | PASS    |                                                                                         |
| `files for input type=email`          | PASS    |                                                                                         |
| `files for input type=password`       | PASS    |                                                                                         |
| `files for input type=date`           | PASS    |                                                                                         |
| `files for input type=month`          | PASS    |                                                                                         |
| `files for input type=week`           | PASS    |                                                                                         |
| `files for input type=time`           | PASS    |                                                                                         |
| `files for input type=datetime-local` | PASS    |                                                                                         |
| `files for input type=number`         | PASS    |                                                                                         |
| `files for input type=range`          | PASS    |                                                                                         |
| `files for input type=color`          | PASS    |                                                                                         |
| `files for input type=checkbox`       | PASS    |                                                                                         |
| `files for input type=radio`          | PASS    |                                                                                         |
| `files for input type=submit`         | PASS    |                                                                                         |
| `files for input type=image`          | PASS    |                                                                                         |
| `files for input type=reset`          | PASS    |                                                                                         |
| `files for input type=button`         | PASS    |                                                                                         |
| `files for input type=file`           | PASS    |                                                                                         |
| `setting <input type=file>.files`     | FAIL    | `assert_equals: expected object "[object FileList]" but got object "[object FileList]"` |

</details>
</details>

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/257217575)

# Sauce (safari) #
Testing web-platform-tests at revision 465b82d6389b220a2c92e72b32fdc4df184208f1
Using browser at version 10.0
Starting 10 test iterations
All results were stable
## All results ##
<details>
<summary>1 test ran</summary>
<details>
<summary><a href="http://w3c-test.org/submissions/6617/html/semantics/forms/the-input-element/files.html">/html/semantics/forms/the-input-element/files.html</a></summary>


|                Subtest                | Results |                            Messages                           |
|---------------------------------------|---------|---------------------------------------------------------------|
|                                       | OK      |                                                               |
| `files for input type=hidden`         | PASS    |                                                               |
| `files for input type=text`           | PASS    |                                                               |
| `files for input type=search`         | PASS    |                                                               |
| `files for input type=tel`            | PASS    |                                                               |
| `files for input type=url`            | PASS    |                                                               |
| `files for input type=email`          | PASS    |                                                               |
| `files for input type=password`       | PASS    |                                                               |
| `files for input type=date`           | PASS    |                                                               |
| `files for input type=month`          | PASS    |                                                               |
| `files for input type=week`           | PASS    |                                                               |
| `files for input type=time`           | PASS    |                                                               |
| `files for input type=datetime-local` | PASS    |                                                               |
| `files for input type=number`         | PASS    |                                                               |
| `files for input type=range`          | PASS    |                                                               |
| `files for input type=color`          | PASS    |                                                               |
| `files for input type=checkbox`       | PASS    |                                                               |
| `files for input type=radio`          | PASS    |                                                               |
| `files for input type=submit`         | PASS    |                                                               |
| `files for input type=image`          | PASS    |                                                               |
| `files for input type=reset`          | PASS    |                                                               |
| `files for input type=button`         | PASS    |                                                               |
| `files for input type=file`           | PASS    |                                                               |
| `setting <input type=file>.files`     | FAIL    | `assert_throws: function "() => i1.files = []" did not throw` |

</details>
</details>

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/257217583)

# Chrome (unstable) #
Testing web-platform-tests at revision 465b82d6389b220a2c92e72b32fdc4df184208f1

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/257217584)

# Sauce (MicrosoftEdge) #
Testing web-platform-tests at revision 465b82d6389b220a2c92e72b32fdc4df184208f1

**Comment by sideshowbarker** [comment]
w3c-test:mirror

File: html/semantics/forms/the-input-element/files.html [hunk 1]
```diff
@@ -45,4 +45,22 @@
...
+  i1.type = "file";
+  i2.type = "file";
+
+  const files = i2.files;
+  i1.files = i2.files;
+  assert_equals(i1.files, files);
+  assert_equals(i2.files, files);
+
+  i1.files = null;
+  assert_equals(i1.files, files);
```
Comment (domenic) [inline html/semantics/forms/the-input-element/files.html]: I think all these different cases should be separate tests, or at the very least should have nice assertion messages.

File: html/semantics/forms/the-input-element/files.html [hunk 2]
```diff
@@ -45,4 +45,22 @@
...
+
+  assert_throws(new TypeError(), () => i1.files = []);
+  assert_throws(new TypeError(), () => i1.files = [new File([], "x")]);
+}, "setting <input type=file>.files");
 </script>
```
Comment (domenic) [inline html/semantics/forms/the-input-element/files.html]: We should also test the "does not apply" setter case.

**Review by domenic — CHANGES_REQUESTED** [review]

**Review by domenic — APPROVED** [review]

**Comment by guest271314** [comment]
@annevk Is the expected result for `change` event to be dispatched when `.files` property is set at `<input type="file">`?

**Pull request merged**